Don't create problematic ownerReferences for CDI and NetworkAddonsConfig

maya-r opened the pull request:
if HCO operator (scope namespace) owns CDI (scope cluster) and
the kubernetes garbage collector finds out, it will delete CDI.

In an update, if an ownerReference exists, remove it.

AsherShoshan commented:
HCO Operator creates the CDI cr, and puts HCO cr as the owner. so the new lines of code is redundant; only the removal of the setownerreference is required.  AFAIK, no one else puts additional owner reference in the CDI kind cr.
But then , when you delete the HCO cr (undeploying CNV), CDI will not be undeployed..

dankenigsberg commented:
@AsherShoshan the added code is not redundant - it is needed when upgrading from an older cluster.

You are right about this patch not handing the deletion use case - it should.

AlonaKaplan commented:
> @AsherShoshan the added code is not redundant - it is needed when upgrading from an older cluster.
> 
> You are right about this patch not handing the deletion use case - it should.

Why do we need to add a special caring for the deletion use case?
Is there a use case for removing the hco cr directly and not due to unregistration from the olm?
If not, I think we don't need the finalizer. When unregistering from the olm, the cdi cr will be removed anyway when the olm will remove the cdi crd.

dankenigsberg commented:
> > You are right about this patch not handing the deletion use case - it should.
> 
> Why do we need to add a special caring for the deletion use case?
> Is there a use case for removing the hco cr directly and not due to unregistration from the olm?
> If not, I think we don't need the finalizer. When unregistering from the olm, the cdi cr will be removed anyway when the olm will remove the cdi crd.

@AlonaKaplan good point. When the cdis.cdi.kubevirt.io is removed, the CDI object would be removed with it even if the latter has no owner.

AsherShoshan commented:
Alona,
There is a UI use case of deleting+redeploying the HCO cr, without removing
all OLM operators
Asher

On Wed, Jan 1, 2020 at 1:36 PM Dan Kenigsberg <notifications@github.com>
wrote:

> You are right about this patch not handing the deletion use case - it
> should.
>
> Why do we need to add a special caring for the deletion use case?
> Is there a use case for removing the hco cr directly and not due to
> unregistration from the olm?
> If not, I think we don't need the finalizer. When unregistering from the
> olm, the cdi cr will be removed anyway when the olm will remove the cdi crd.
>
> @AlonaKaplan <https://github.com/AlonaKaplan> good point. When the
> cdis.cdi.kubevirt.io is removed, the CDI object would be removed with it
> even if the latter has no owner.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393?email_source=notifications&email_token=ALCEMQK4VXLPFKKKC3ZEVQ3Q3R54ZA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5DNGQ#issuecomment-570046106>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCEMQJPQMSHGVXTYDCYRATQ3R54ZANCNFSM4KBVZOBQ>
> .
>

AlonaKaplan commented:
> Alona, There is a UI use case of deleting+redeploying the HCO cr, without removing all OLM operators Asher

@AsherShoshan can you please elaborate? Do you mean there is an option from the ui to remove the hco cr? Why would the user want to do it? Maybe we should block it.


> […](#)
> On Wed, Jan 1, 2020 at 1:36 PM Dan Kenigsberg ***@***.***> wrote: You are right about this patch not handing the deletion use case - it should. Why do we need to add a special caring for the deletion use case? Is there a use case for removing the hco cr directly and not due to unregistration from the olm? If not, I think we don't need the finalizer. When unregistering from the olm, the cdi cr will be removed anyway when the olm will remove the cdi crd. @AlonaKaplan <https://github.com/AlonaKaplan> good point. When the cdis.cdi.kubevirt.io is removed, the CDI object would be removed with it even if the latter has no owner. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#393?email_source=notifications&email_token=ALCEMQK4VXLPFKKKC3ZEVQ3Q3R54ZA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5DNGQ#issuecomment-570046106>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ALCEMQJPQMSHGVXTYDCYRATQ3R54ZANCNFSM4KBVZOBQ> .

AsherShoshan commented:
Yes.
After operator is installed (vis operator hub page).
You can remove/create hco.
More than that - in UI you can explicitly create each cr (kubevirt, cdi,
etc).. without being owned by HCO cr.

On Wed, Jan 1, 2020 at 3:25 PM AlonaKaplan <notifications@github.com> wrote:

> Alona, There is a UI use case of deleting+redeploying the HCO cr, without
> removing all OLM operators Asher
>
> @AsherShoshan <https://github.com/AsherShoshan> can you please elaborate?
> Do you mean there is an option from the ui to remove the hco cr? Why would
> the user want to do it? Maybe we should block it.
>
> … <#m_-14441822535483444_>
> On Wed, Jan 1, 2020 at 1:36 PM Dan Kenigsberg *@*.***> wrote: You are
> right about this patch not handing the deletion use case - it should. Why
> do we need to add a special caring for the deletion use case? Is there a
> use case for removing the hco cr directly and not due to unregistration
> from the olm? If not, I think we don't need the finalizer. When
> unregistering from the olm, the cdi cr will be removed anyway when the olm
> will remove the cdi crd. @AlonaKaplan <https://github.com/AlonaKaplan>
> https://github.com/AlonaKaplan good point. When the cdis.cdi.kubevirt.io
> is removed, the CDI object would be removed with it even if the latter has
> no owner. — You are receiving this because you were mentioned. Reply to
> this email directly, view it on GitHub <#393
> <https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393>?email_source=notifications&email_token=ALCEMQK4VXLPFKKKC3ZEVQ3Q3R54ZA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5DNGQ#issuecomment-570046106>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/ALCEMQJPQMSHGVXTYDCYRATQ3R54ZANCNFSM4KBVZOBQ
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393?email_source=notifications&email_token=ALCEMQKXIMSJU4K55G2GESTQ3SKVBA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5FDJQ#issuecomment-570053030>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCEMQOSCUN3TBJRYZWKBP3Q3SKVBANCNFSM4KBVZOBQ>
> .
>

dankenigsberg commented:
Sorry @AsherShoshan, I do not get your point. How is editing various CRs related to our discussion?

AsherShoshan commented:
Dan,
I didn't write editing...
In UI, you can delete the kind-HCO cr (without removing OLM operators), and
recreate it.
Also you can, for example, create kind-CDI cr - without going via kind-HCO
cr.

On Wed, Jan 1, 2020 at 5:00 PM Dan Kenigsberg <notifications@github.com>
wrote:

> Sorry @AsherShoshan <https://github.com/AsherShoshan>, I do not get your
> point. How is editing various CRs related to our discussion?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393?email_source=notifications&email_token=ALCEMQIOVHB5MYKA54FTWFTQ3SVZHA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5GXUY#issuecomment-570059731>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCEMQM3R6DBVAV6NXCE7CTQ3SVZHANCNFSM4KBVZOBQ>
> .
>

dankenigsberg commented:
@AsherShoshan what you are saying is not clear to me. What do you expect to happen when the kind-hco rc is deleted? Does it happen now? Would it happen with this PR?

AsherShoshan commented:
On Wed, Jan 1, 2020 at 10:18 PM Dan Kenigsberg <notifications@github.com>
wrote:

> @AsherShoshan <https://github.com/AsherShoshan> what you are saying is
> not clear to me. What do you expect to happen when the kind-hco rc is
> deleted? Does it happen now? Would it happen with this PR?
>
I replied Alona regarding a use case of HCO cr deletion without HCO CRD
deletion.
In UI, you can delete only the HCO-kind cr, then all dependent cr are
deleted, but OLM operators are not deleted.
Also, in UI, when only OLM operators installed (i.e kubevirt-hyperconverged
operator installed via OperatorHub page), you can create each cr (kubevirt,
cdi, etc), without creating the HCO cr.  It means kubevirt-kind has no
owner.

IIUC, after merging the PR, if you delete the HCO cr via UI (without OLM
removal), dependant cr's will not be deleted, and VM, VMI will not be
deleted.
with the second use case, with the fix of Omer, (to put HCO CRD as owner of
VM,VMI CRD), deleting kubevirt-kind cr will not delete VM, VMI either.

You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubevirt/hyperconverged-cluster-operator/pull/393?email_source=notifications&email_token=ALCEMQKTEWIEW74LN7E2UILQ3T3ARA5CNFSM4KBVZOB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH5L24I#issuecomment-570080625>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALCEMQJCS2JDULG2HOMH45DQ3T3ARANCNFSM4KBVZOBQ>
> .
>

dankenigsberg commented:
Yes, this pr changes the behavior if kind-hco is deleted (either by ui or
command line, it does not matter).

However, I don't think that deleting kind-hco cr is a valid use case. It
was not an acceptable method to cleanly remove hco before this pr, and
neither after.

rthallisey commented:
I don't think it's a good idea to change the documentation behavior of cleanup without much notice.  There's been some discussion of changing the behavior in the past, but there should be a formal agreement from all (PM & ENG) and a controlled rollout, with plenty of time for testing, of the change.  In the near term, the underlying problem of garbage collection should be the priority and can be fixed without changing the cleanup behavior.   See https://github.com/kubevirt/hyperconverged-cluster-operator/pull/127/files#diff-64e1e2b9251fa5137c91d5a7c8427b88R35.  Long term, it would be worth discussing cleanup.

dankenigsberg commented:
@rthallisey I am not aware of any change to any documented behavior incurred by this PR. Could you point me to any?

maya-r commented:
Asking for clarification, there are two kinds of cleanup:
- Cleanup on deletion, discussed in the replies, not done in the PR
- Cleanup on update, which I've attempted in the PR

Would you prefer not to cleanup on update yet, too?

rthallisey commented:
>Asking for clarification, there are two kinds of cleanup:
>
>    Cleanup on deletion, discussed in the replies, not done in the PR
>
Cleanup of CDI's CR when HCO's CR is deleted will no longer happen after this PR is merged.  I'm saying that this changes the expectation for cleanup and that I don't think that's the right approach.  IMO there should be a discussion before changing how things are cleaned up, so we don't end up bundling a feature together with a bug fix.

I think we should look if it's still possible to solve with finalizes as @phoracek did here ( https://github.com/kubevirt/hyperconverged-cluster-operator/pull/127/files#diff-64e1e2b9251fa5137c91d5a7c8427b88R35)

Docs: https://docs.openshift.com/container-platform/4.2/cnv/cnv_install/uninstalling-container-native-virtualization.html 

>    Cleanup on update, which I've attempted in the PR
>
>Would you prefer not to cleanup on update yet, too?
>

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -492,6 +489,18 @@ func (r *ReconcileHyperConverged) ensureCDI(instance *hcov1alpha1.HyperConverged
+		logger.Info("CDI has owners, removing...")
+		noOwnerRefs := []metav1.OwnerReference{}
+		found.SetOwnerReferences(noOwnerRefs)
+		r.client.Update(context.TODO(), found)
```
Please handle the error.

djzager replied:
```suggestion
		return r.client.Update(context.TODO(), found)
```

Seems sufficient to me.

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -492,6 +489,18 @@ func (r *ReconcileHyperConverged) ensureCDI(instance *hcov1alpha1.HyperConverged
+	// It's not legal, so remove that.
+	if (len(existingOwners) > 0) {
+		logger.Info("CDI has owners, removing...")
+		noOwnerRefs := []metav1.OwnerReference{}
```
```suggestion
```

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -492,6 +489,18 @@ func (r *ReconcileHyperConverged) ensureCDI(instance *hcov1alpha1.HyperConverged
+	if (len(existingOwners) > 0) {
+		logger.Info("CDI has owners, removing...")
+		noOwnerRefs := []metav1.OwnerReference{}
+		found.SetOwnerReferences(noOwnerRefs)
```
```suggestion
		found.SetOwnerReferences([]metav1.OwnerReference{})
```

dankenigsberg commented:
@rthallisey https://docs.openshift.com/container-platform/4.2/cnv/cnv_install/uninstalling-container-native-virtualization.html#cnv-deleting-kubevirt-hyperconverged-custom-resource_uninstalling-container-native-virtualization says that one one remove the hco.cr, but I am not completely sure why. Wouldn't the next step https://docs.openshift.com/container-platform/4.2/cnv/cnv_install/uninstalling-container-native-virtualization.html#cnv-deleting-hco-subscription_uninstalling-container-native-virtualization trigger the removal of cnv's CRDs, and by that the removal of cdi.cr ?

However, since HCO is the one creating cdi.cr, it makes sense that HCO would also be the one to delete it. @maya-r would you add a finalizer to remove cdi.cr on hco.cr deletion?

dankenigsberg commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -470,9 +470,6 @@ func newCDIForCR(cr *hcov1alpha1.HyperConverged, namespace string) *cdiv1alpha1.
 
 func (r *ReconcileHyperConverged) ensureCDI(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request) error {
 	cdi := newCDIForCR(instance, UndefinedNamespace)
-	if err := controllerutil.SetControllerReference(instance, cdi, r.scheme); err != nil {
```
line 89 in this file depends on CDI being owned by the hco.cr. With the suggested code, HCO would not be notified if someone edits CDI.cr. For completeness, please make this controller watch the cdi.cr regardless of its owner.

maya-r commented:
Merged chunks of #380 and #122

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -601,6 +675,18 @@ func (r *ReconcileHyperConverged) ensureNetworkAddons(instance *hcov1alpha1.Hype
 	err = r.client.Get(context.TODO(), key, found)
+
+	// Handle HCO finalizer
+	if !instance.ObjectMeta.DeletionTimestamp.IsZero() {
```
Should go away and be centrally located in the main `Reconcile()`

coypoop commented:
Is this roughly what you had in mind? https://github.com/kubevirt/hyperconverged-cluster-operator/compare/master...maya-r:cdi-owner6 (having some issues with setting it up while at home, so not tested).

dankenigsberg commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -601,6 +675,18 @@ func (r *ReconcileHyperConverged) ensureNetworkAddons(instance *hcov1alpha1.Hype
+	if !instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		errRemoval := manageComponentResourceRemoval(found, r.client, instance)
+		if errRemoval != nil {
+			logger.Error(err, "Failed during CDI removal")
```
wrong error string?

maya-r commented:
This resembles the existing reconcile logic, hopefully this is your intention.
In the long run I feel compelled to have a "getCDI" that gets the actual CDI instance.

ormergi commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -601,13 +649,28 @@ func (r *ReconcileHyperConverged) ensureNetworkAddons(instance *hcov1alpha1.Hype
+	// as the owner of NetworkAddons (scope cluster).
+	// It's not legal, so remove that.
+	if (len(existingOwners) > 0) {
+		logger.Info("NetworkAddons has owners, removing...")
```
consider clearer log message for example: Removing NetworkAddonsConfigs owner references

ormergi commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -187,6 +199,33 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+			return reconcile.Result{}, r.client.Update(context.TODO(), instance)
+		}
+	} else {
+		for _, f := range []func(c client.Client, cr *hcov1alpha1.HyperConverged) error{
```
I think there should be another check for finalizer existence in case something went wrong and the finalizer is no set.

djzager replied:
seems reasonable.

djzager replied:
There is some good information in the operator-sdk documentation regarding [finalizers](https://github.com/operator-framework/operator-sdk/blob/master/doc/user-guide.md#handle-cleanup-on-deletion).

Checking for the finalizer and not removing the finalizer unless the cleanup proceeds successfully sounds like a good idea.

ormergi commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -187,6 +199,33 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+
+		// Remove the finalizer
+                instance.ObjectMeta.Finalizers = drop(instance.ObjectMeta.Finalizers, FinalizerName)
+
```
Please add foreground finalizer removal in case there is one, so it wont block the HCO deletion process:
```
if contains(instance.ObjectMeta.Finalizers, foregroundDeletionFinalizer) &&
    len(instance.ObjectMeta.Finalizers) == 1 {
    instance.ObjectMeta.Finalizers = drop(instance.ObjectMeta.Finalizers, foregroundDeletionFinalizer)
}
```

djzager replied:
are we even using the foregroundDeletionFinalizer? I don't see anywhere where we would be adding that finalizer to a HyperConverged resource.

If not, I would prefer us not adding extra logic to cleanup a finalizer we know we aren't adding.

ormergi replied:
I searched in hco, kubevit, cnao, cdi, ssp operators repository's and there is  no usage of foreground deletion finalizer for HCO CR, so its looks like its safe not to add it.

ormergi commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -1085,3 +1148,112 @@ func isKVMAvailable() bool {
 }
+
+// GetNamespacedName returns the name/namespace of the HyperConverged resource
+func GetNamespacedName() (types.NamespacedName, error) {
```
I think this function should be private

ormergi commented:
Please consider those changes:
1. Refactoring the functions GetNamespacedName, contains, drop, toUnstructured, 
componentResourceRemoval. to be more readable (i.e: meaningful variables names) and more testable (i.e: in GetNamespacedName the function using external variable instead of pass it as argument).

2. I think GetNamespacedName should be private 

3. At reconcile function when handling finalizers in case there is a deletionTimeStamp add another check for finalizer existence in case something went wrong and the finalizer is no set.

4. At reconcile function when handling finalizers, when removing the finalizer please add handling for removing foregroundDeletionFinalizer.

maya-r commented:
>     2. I think GetNamespacedName should be private

It isn't private because it's taken from #380 where it's used as a public function. I picked a minimal part of #380 to make this pull request work, trying not to conflict with further development there and make sure this pull request can be merged whether #380 is merged before it or not.

A name like "GetHyperconverged" would be good, but I think it makes more sense to make this change in #380

>     3. At reconcile function when handling finalizers in case there is a deletionTimeStamp add another check for finalizer existence in case something went wrong and the finalizer is no set.
> 
>     4. At reconcile function when handling finalizers, when removing the finalizer please add handling for removing foregroundDeletionFinalizer.

I don't think we should remove any finalizers we are not currently adding, someone may have their own use case for adding finalizers on top of of ours, and we'll be fighting them.

dankenigsberg commented:
> > ```
> > 2. I think GetNamespacedName should be private
> > ```
> 
> It isn't private because it's taken from #380 where it's used as a public function. I picked a minimal part of #380 to make this pull request work, trying not to conflict with further development there and make sure this pull request can be merged whether #380 is merged before it or not.

Each PR should be self-contained. Right now we do not know if #380 is to be merged nor when. Therefore I would vote to keep the function private. When it has to be exposed, this can easily be done.

rthallisey commented:
/test hco-e2e-aws

maya-r commented:
I couldn't think of better names for the variables in 'toUnstructured', suggestions welcome.
I think it's very important to have a working fix in-tree so we can test it and become happy by the expected release time. we can create additional pull requests for additional cleanup later.

djzager commented:
> Each PR should be self-contained. Right now we do not know if #380 is to be merged nor when. Therefore I would vote to keep the function private. When it has to be exposed, this can easily be done.

I agree. Go ahead and make it private.

djzager commented:
You've changed the way the Reconcile function works. You will need to update the tests to reflect these changes.

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -187,6 +199,33 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+                instance.ObjectMeta.Finalizers = drop(instance.ObjectMeta.Finalizers, FinalizerName)
+
+                // Need to requeue because finalizer update does not change metadata.generation
+                return reconcile.Result{}, r.client.Update(context.TODO(), instance)
```
need to set `Requeue: true` if you actually want to force requeu.

maya-r commented:
What's a way of running the tests locally? `make test-unit` doesn't have the intended effect.

djzager commented:
> What's a way of running the tests locally? `make test-unit` doesn't have the intended effect.

`make test`

djzager commented:
I see there is some wonkiness between `make test` and `make test-unit` ... odd ... but I don't think now is the time to play around with the Makefile.

djzager commented:
/retest

djzager commented:
I think code coverage (coveralls) should be taken up in a follow-up PR. Focus on getting the existing tests to pass.

maya-r commented:
Changes since previous version:
- Fix whitespace
- Combine commits into one (rather than create a function and immediately modify it)
- Only attempt to remove the finalizer if it exists, suggested by ormergi
- Don't requeue another reconcile after installing the finalizer, there's no problem
  with just continuing our current run, and it upsets the tests that expect a certain
  state after the first reconcile.

ormergi commented:
lgtm

rthallisey commented:
ci test please

orenc1 commented:
@maya-r please also handle ownerReference removal (& cleanup at finalizer) of KubevirtCommonTemplatesBundle, whose owner/creator is also the HCO.

maya-r commented:
> @maya-r please also handle ownerReference removal (& cleanup at finalizer) of KubevirtCommonTemplatesBundle, whose owner/creator is also the HCO.

done

djzager commented:
@maya-r Wow. That stinks. All green except for code coverage and I'm asking for a change.

I still think you need to handle the err for getting the kubevirt common templates bundle before handing the owner reference stuff. Everything else looks great.

tiraboschi commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -41,9 +45,7 @@ const (
-	// Foreground deletion finalizer is blocking removal of HyperConverged until explicitly dropped.
-	// TODO: Research whether there is a better way.
-	foregroundDeletionFinalizer = "foregroundDeletion"
+	HyperConvergedName = "hyperconverged-cluster"
```
I think it will be better to let the user express a custom value with an environmental variable keeping this as a sane default.

djzager replied:
my reason for not pursuing this route is that it's going to be backed into the CSV already so the user will not be able to configure it when installing via OLM. Why not remove the variation and simply encode the value in the source?

tiraboschi replied:
On the other side we are going to fail if, by mistake, the user creates the CR with a different name.
We can set consume this a default, set an override via env variable in the CSV and compose the alm-example in the CSV according to what we have in the environment variable.
See: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/399

djzager replied:
Why not just put the "hyperconverged-cluster" in the alm-example, done. How does using an environment variable solve the problem of

> On the other side we are going to fail if, by mistake, the user creates the CR with a different name.

This will still be a problem if we use an environment variable. The facts are, regardless of hardcoded in source or using an environment variable, that:

1. The user must create the resource in the same namespace as the operator
1. The name of that resource is determined long before the user "Subscribes" to the operator

This was the reason @dankenigsberg that I wanted to specifically handle this scenario where a user creates a HyperConverged resource that we weren't expecting, either:

1. They mistakenly created the resource with the wrong name or namespace. 
1. They mistakenly created the resource when one we are reconciling already exists.

In either case, the polite thing to do is notify the user (via status), "Hey, this CR doesn't look right."

dankenigsberg replied:
Found it awkward to touch wrongly created CR. But when thinking about it again I suspect it would be ok for the hyperconverged controller to track all hyperconverged CRs.

Anyway, this should be discussed on another PR.

maya-r commented:
> @maya-r Wow. That stinks. All green except for code coverage and I'm asking for a change.
> 
> I still think you need to handle the err for getting the kubevirt common templates bundle before handing the owner reference stuff. Everything else looks great.

thanks, fixed!

djzager commented:
/retest

djzager commented:
with only code coverage remaining. I'm accepting this PR.

tiraboschi commented:
/cherrypick release-2.2
